File: 7_neet_code_150_problems/1_foundation_1_2_weeks/2_two_pointers/valid_palindrome_.go
package main

import (
	"fmt"
	"unicode"
)

func cleanupString(s string) string {
	result := make([]rune, 0, len(s))

	for _, r := range s {
		if unicode.IsLetter(r) || unicode.IsNumber(r) {
			result = append(result, unicode.ToLower(r))
		}
	}
	return string(result)
}

func isValidPalindrome(s string) bool {
	cStr := cleanupString(s)
	i := 0
	j := len(cStr) - 1
	for i < j {
		if cStr[i] != cStr[j] {
			return false
		}
		i++
		j--
	}
	return true
}

func main() {
	str_ := "Was it a car or a cat I saw?"
	fmt.Println(isValidPalindrome(str_))
}

File: 7_neet_code_150_problems/1_foundation_1_2_weeks/2_two_pointers/valid_palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = "".join([x for x in list(s) if x.isalnum()])
        i, j = 0, len(s) - 1
        print(s)
        while i < j:
            if s[i].lower() != s[j].lower():
                return False
            i += 1
            j -= 1
        return True



s = Solution()
str_ = "Was it a car or a cat I saw?"
print(s.isPalindrome(str_))